File: 0930-binary-subarrays-with-sum/0930-binary-subarrays-with-sum.cs
public class Solution {
    public int NumSubarraysWithSum(int[] nums, int goal) {
        Dictionary<int, int> cumulativeSumCount = new Dictionary<int, int>();
        cumulativeSumCount[0] = 1;  // Initialize with sum 0 occurring once
        int cumulativeSum = 0;
        int count = 0;

        foreach (int num in nums) {
            cumulativeSum += num;

            if (cumulativeSumCount.ContainsKey(cumulativeSum - goal)) {
                count += cumulativeSumCount[cumulativeSum - goal];
            }

            if (cumulativeSumCount.ContainsKey(cumulativeSum)) {
                cumulativeSumCount[cumulativeSum]++;
            } else {
                cumulativeSumCount[cumulativeSum] = 1;
            }
        }

        return count;
    }
}

File: 0930-binary-subarrays-with-sum/0930-binary-subarrays-with-sum.cpp
class Solution {
public:
    int numSubarraysWithSum(vector<int>& nums, int goal) {
        unordered_map<int,int>ump;
        ump[0] = 1;
        int n = nums.size();
        int countNumber = 0;
        int cumulativeSum = 0;
        for(int i=0; i<n; i++)
        {
            cumulativeSum+=nums[i];
            if(ump.find(cumulativeSum-goal)!=ump.end())
            {
                countNumber += ump[cumulativeSum-goal];
            }
            ump[cumulativeSum]++;
        }
        return countNumber;
    }
};